java/src/test/java/org/extism/sdk/PluginTests.java

```diff
@@ -104,4 +110,44 @@ public void shouldAllowInvokeFunctionFromFileWasmSourceMultipleTimesByReusingCon
         }
     }
 
+    @Test
+    public void shouldAllowInvokeHostFunctionFromPDK() {
+        var parametersTypes = new LibExtism.ExtismValType[]{LibExtism.ExtismValType.I64};
+        var resultsTypes = new LibExtism.ExtismValType[]{LibExtism.ExtismValType.I64};
+
+        ExtismFunction helloWorldFunction = (ExtismCurrentPlugin plugin, LibExtism.ExtismVal[] params, LibExtism.ExtismVal[] returns, Optional<JsonElement> data) -> {
+            System.out.println("Hello from Java Host Function!");
+            System.out.println(String.format("Input string received from plugin, %s", plugin.inputString(params[0])));
+
+            int offs = plugin.alloc(4);
+            Pointer mem = plugin.memory();
+            mem.write(offs, data.map(JsonElement::getAsString).orElse("test").getBytes(), 0, 4);
+            returns[0].v.i64 = offs;
+        };
+
+        String str = "test";
+        Pointer hostUserData = new Memory(str.length() + 1);
```

Comment:
I think there is a slight misunderstanding about the user data here. it is not something the plugin typically accesses. You can think of it like applied arguments in a closure, or some properties in a class. In this case it's for the java code, and not the plugin  code.

Here is an example of passing in a db connection: https://github.com/extism/extism-sqlite-host-function

here you could pass in things specific to the user. E.G. a session or something.

I'm taking a look at strategies to handle this in Java today.

Reply:
I'm a bit unsure how to make all this work just trying to express the idea right now.

Reply:
With other languages, like python, we have something like a concept of handles: https://github.com/extism/extism/blob/main/python/extism/extism.py#LL205C30-L205C45

this allows us to pass a handle to python objects across the ffi boundary, then when we come back we can retried those python objects from the handle. We need the equivalent thing. I asked ChatGPT what it thinks the equivalent is and it said Callback https://java-native-access.github.io/jna/4.2.1/com/sun/jna/Callback.html

so perhaps we can put some data in a callback and pull it out when we get back into java world?